modules/generator/processor/servicegraphs/servicegraphs.go

```diff
@@ -301,12 +324,22 @@ func (p *Processor) onExpire(e *store.Edge) {
 		// We check if the span we have is the root span, and if so, we set the client service to "user".
 		if _, parentSpan := parseKey(e.Key()); len(parentSpan) == 0 {
 			e.ClientService = "user"
+
+			if p.Cfg.EnableVirtualNodeLabel {
+				e.Dimensions["virtual_node"] = "client"
+			}
+
 			p.onComplete(e)
 		}
 	} else if len(e.ServerService) == 0 && len(e.PeerNode) > 0 {
 		// If client span does not have its matching server span, but has a peer attribute present,
 		// we make the assumption that a call was made to an external service, for which Tempo won't receive spans.
 		e.ServerService = e.PeerNode
+
+		if p.Cfg.EnableVirtualNodeLabel {
+			e.Dimensions["virtual_node"] = "server"
+		}
+
```

Comment:
`onExpire` needs to call `ReturnEdge` when it doesn't call onComplete.

Reply:
adding this - it doesn't seem to have coverage though, so I'll look into that too soon